cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/constraint/DelayConstraint.java

```diff
@@ -16,16 +16,33 @@
 
 package co.cask.cdap.internal.app.runtime.schedule.constraint;
 
-import co.cask.cdap.internal.schedule.constraint.Constraint;
+import co.cask.cdap.internal.app.runtime.schedule.ProgramSchedule;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
 
 /**
  * A constraint which requires a certain amount of delay pass after the trigger is fired, before executing the program.
  */
-public class DelayConstraint extends Constraint {
+public class DelayConstraint extends AbstractCheckableConstraint {
 
   private final long millisAfterTrigger;
 
   public DelayConstraint(long millisAfterTrigger) {
+    Preconditions.checkArgument(millisAfterTrigger >= 0,
+                                "Configured delay '%s' must not be a negative value.", millisAfterTrigger);
     this.millisAfterTrigger = millisAfterTrigger;
   }
+
+  public ConstraintResult check(ProgramSchedule schedule, ConstraintContext context) {
+    return check(System.currentTimeMillis(), context.getJob().getCreationTime());
```

Comment:
Better have the "checkTime" exposed from the `ConstraintContext` and have constraint check to use that instead of `System.currentTimeMillis`. This will make testing easier and we can provide the same check time for all constraints in the same job for consistency.